### ISSUE TITLE
Allow baking feature flags into a release for self-hosted

### ISSUE THREAD

**Issue by macobo** [PostHog/posthog#2141]
## Is your feature request related to a problem?

Feature flags are awesome, feature flags are great. We use feature flags as part of posthog development as well.

However due to how our product works, this can have some downsides, where enabling a feature in cloud may enable a (unfinished version of the feature) on clients self-hosted instances

## Describe the solution you'd like

Update frontend/src/lib/logic/featureFlagLogic.ts to have a hard-coded list of features when on self-hosted instances (and not in DEBUG/SELF-REPORT).

## Describe alternatives you've considered

Not using feature flags for developing posthog. This misses a great opportunity though.

## Additional context

Related: https://github.com/PostHog/posthog/issues/1996

**Comment by paolodamico** [comment]
I'm not sure about this, a lot of our usage comes from self-hosted, and we're starting to run more and more experiments to test the impact of changes before making them generally available. We would be limiting our experimentation abilities considerably by hard-coding this. Also, we could just enable feature flags based on a user cohort that only groups cloud users. Thoughts?

**Comment by macobo** [comment]
If we still wanted to do experiments, we could bake in release percentage and use the unique installation ID to decide whether the flag should be on or not.

In my mind it's really important that we don't accidentally break old installations - accidentally releasing or removing half-baked features definetly counts towards bugginess of the platform.

**Comment by mariusandra** [comment]
I do think this proposal makes sense. Releases should be "stable". I.e. they shouldn't change unexpectedly.

Several times I've heard users say things like "I didn't update my posthog version, but X and Y changed totally. What's happening?", with the added implication that we're using some dark "phone home" magic to worm into their VPN and flip some switches.

New problem: What if a flag is there, but disabled for 3 consecutive posthog versions? Then once we finally fix all the bugs on `master`, we turn on the flag, including for users that are 2 versions behind. I think this has already happened, yet we just don't hear about it unless things just start breaking.

In short, there are too many variables here to provide a good and consistent #experience if we enable/disable flags at will for self-hosted users. Hence it might make sense to at least commit a hard-coded "leave these flags at this state" list when making a release. Even if we enable support for additional flags on top.

**Comment by paolodamico** [comment]
Yeah I guess this makes sense. It'll also help us have a better log of when features were fully released on self-hosted. Not sure if we should do %-based rollouts on machine ID as a self-hosted installation could have multiple nodes and this could lead to a wildly disparate experience, but also not a big deal. We can test this on cloud.

**Comment by macobo** [comment]
> Not sure if we should do %-based rollouts on machine ID as a self-hosted installation could have multiple nodes

Yeah, it's not ideal. If we wanted to be cheeky we could use SECRET_KEY as the thing to hash and decide rollout based on - I'd hope it's unique in all installs :)

**Comment by paolodamico** [comment]
Alright here's my proposal for the static flags on the next release cycle. I'm assuming we'll do hashing based on the `SECRET_KEY`, but it's also fine if we do only binary on/off to avoid any weird behavior. I'm excluding things that have already been released from feature flags in this release (e.g. new query experience).

```
papercups-enabled - OFF 
new-onboarding-2822 - OFF (experiment was unsuccessful, we're dropping the feature the way it currently is)
save-cohort-on-modal - OFF (** checking with author if we can continue release)
ingestion-grid-exp-3 - ON (believe results here were positive)
3638-trailing-wau-mau - OFF (has only been released to beta users)
project-home-exp-5 - OFF (believe we want to optimize this before continuing)
4141-event-columns - 50% or OFF (we need to test this more on self-hosted too, but it's definitely not prime time ready).
4562-nps - ON (useful to get input from self-hosted users)
4267-event-property-taxonomy - ON (already released to all users and we haven't had any major bug reports)
4156-tooltips-legends - 50% or OFF (we still need to test more, but getting more eyeballs would be very useful)
4535-funnel-bar-viz - OFF (still WIP)
4819-persons-modal-updates - OFF (still WIP)
```

Tagging @kpthatsme for feedback on `ingestion-grid-exp-3` & `project-home-exp-5`.

**Comment by paolodamico** [comment]
One final thing to consider when implementing this is that we'll want to have a way for users to override this on their end, because there's for instance users that want to test beta features on their self-hosted instances. there should be a way for us to tell them how to enable those beta features on their instance.

One option would be to also ship with a static list of beta features and then with an env var you can enable those.

**Comment by kpthatsme** [comment]
> Alright here's my proposal for the static flags on the next release cycle. I'm assuming we'll do hashing based on the `SECRET_KEY`, but it's also fine if we do only binary on/off to avoid any weird behavior. I'm excluding things that have already been released from feature flags in this release (e.g. new query experience).
> 
> ```
> papercups-enabled - OFF 
> new-onboarding-2822 - OFF (experiment was unsuccessful, we're dropping the feature the way it currently is)
> save-cohort-on-modal - OFF (** checking with author if we can continue release)
> ingestion-grid-exp-3 - ON (believe results here were positive)
> 3638-trailing-wau-mau - OFF (has only been released to beta users)
> project-home-exp-5 - OFF (believe we want to optimize this before continuing)
> 4141-event-columns - 50% or OFF (we need to test this more on self-hosted too, but it's definitely not prime time ready).
> 4562-nps - ON (useful to get input from self-hosted users)
> 4267-event-property-taxonomy - ON (already released to all users and we haven't had any major bug reports)
> 4156-tooltips-legends - 50% or OFF (we still need to test more, but getting more eyeballs would be very useful)
> 4535-funnel-bar-viz - OFF (still WIP)
> 4819-persons-modal-updates - OFF (still WIP)
> ```
> 
> Tagging @kpthatsme for feedback on `ingestion-grid-exp-3` & `project-home-exp-5`.

List LGTM. 

Just want to call out one thing – aside from the rollout, the ability to disable a bad experiment or broken feature is pretty useful. We're going to lose that ability with this change, so we'll just need to be mindful of losing that ability when pushing new self-host releases. 

Usually I've seen two kinds of problems with feature flags
1) Bugs - Usually we can avoid that by only reading from flags where the code is available and writing careful code. 
2) UX changing on a user – not necessarily a bug, might be something about the experiment or how we rolled it out. 

Could you go into a little more around what happened with the cloud vs self-host flag? I wonder if we can solve this problem without switching to a static list.

**Comment by paolodamico** [comment]
Re @kpthatsme. Thanks for reviewing this! @macobo and I talked about it and we'll have some way to override the FF on the static list for edge cases where something goes wrong. 

There's more context on the original issue as to why the static list, but in a nutshell: it's problematic to switch FF values within releases as it can lead to unpredictable and unexpected behavior.

**Comment by mariusandra** [comment]
Today I'm again just not getting the `4156-tooltips-legends` flag on app, and hence all UI tooltips are in the old style. I'm sure if we retroactively disable experiments since they're "released", various users will have similar things happen to them.

Worse, if we enable a flag, yet you're running an old version where the feature behind the flag is just partially developed, you're in for bad times.

Thus, have we agreed we will burn in a list of flags to the next release? What will be the list?

**Comment by paolodamico** [comment]
Closing this as #5273 already does this
